### PR TITLE
IPVGO: Fix linting errors from the new lint rules. 

### PR DESCRIPTION
**Pull request by ficocelliguy**

### CODE REVIEW

**Comment by catloversg** [comment]
@d0sboots Can you review this PR? After you approve these changes, I'll use it as the base commit for a new PR.

You can ignore the change in `.eslintrc.js`. I'll exclude that change.

File: src/Go/boardAnalysis/boardAnalysis.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/boardAnalysis.ts:689]: Did you test this? I don't think it works.

`for...in` iterates over the enumerable string keys of an object. In the case of arrays, that comes out to the indices, although as strings instead of numbers, which is awkward and why you're doing awkward things below.
Both `for...of` and `.forEach` iterate over the *values* of an array (or other collection). What you have in "rowIndexString" now is actually the array slices, not indices.

File: src/Go/boardAnalysis/boardAnalysis.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/boardAnalysis.ts]:
```suggestion
  for (const row of Go.currentGame.board) {
```

File: src/Go/boardAnalysis/boardAnalysis.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/boardAnalysis.ts]:
```suggestion
    for (const point of row) {
      if (point == null) continue;
```
And then you can dispense with all the checks against point, below.

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: These are functional changes, not lint changes. I don't think they belong in this PR?
(I'm also not sure I agree with them in general; I don't think sprinkling sleeps through the code like this is a good idea.)

File: test/jest/Go/boardState.test.ts
Comment (d0sboots) [inline test/jest/Go/boardState.test.ts]: Can't you use `as` here instead of disabling the lint?

**Review by d0sboots — COMMENTED** [review]
Sorry for being so slow here. I chose a bad time to get busy.

File: src/Go/boardAnalysis/boardAnalysis.ts
Comment (ficocelliguy) [reply src/Go/boardAnalysis/boardAnalysis.ts:689]: I need both the index and the value for identifying the coordinates as well as the current color. The original way to do it was clunky, and the first change to it from earlier in this PR did not work. I have cleaned it to use .entries() to get both index and value.

File: test/jest/Go/boardState.test.ts
Comment (ficocelliguy) [reply test/jest/Go/boardState.test.ts]: I changed this to not need the .any() here and better test the actual value of result.board in the next expect()

File: src/Go/boardAnalysis/goAI.ts
Comment (ficocelliguy) [reply src/Go/boardAnalysis/goAI.ts]: I removed these short sleeps and instead just overrode the lint complaint that these methods were async without strictly needing to be.

removing the async from the shorter move lookup methods would require some very big type refactoring in the way I do lazy move type evaluation and caching for performance.

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: This doesn't make sense. You're disabling the lint warning for a useless async, but you're also *adding* `async` to this function in this change. Obviously, it worked fine without `async` before, so you should be able to revert this whole thing and lint shouldn't complain?

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: Same

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: Same

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: Same

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: Sameish (the async here is only the result of the needless async added elsewhere)

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts]: Same

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts:788]: These really shouldn't be async in the first place. The correct fix is to fix that. If you don't want to fix the typing, the next-most-preferable fix is to just disable the lint here.

I see two possible ways to fix the typing that are hopefully easy, either extend the explicit type up above, or (better) remove the explicit type entirely, and put `const` at the *end* of this object literal definition. That will cause TS to infer the record type specifically and directly from the members. 

An aside: because of the syntax, it's easy to think that making a function `async` is a simple change. However, the reality is that it's a very *big* change, with significant (but subtle) implications for state consistency and invariants. Ideally, very few of these functions would be async.

Comment (ficocelliguy) [reply src/Go/boardAnalysis/goAI.ts:788]: All of the move option getters are awaited when they are referenced abstractly on line 810; previously having them all return a promise made typing much cleaner, since they are intended to be interchangeable.

I've refactored all of these to limit the async to just the few that are particularly slow (and interact with bonus time). This results in a significant amount of code change, but in my testing everything seems to work as expected after this refactor.

File: src/Go/boardAnalysis/goAI.ts
Comment (d0sboots) [inline src/Go/boardAnalysis/goAI.ts:767]: I'm pretty sure that you can leave off declaring the type here, and you might not need `MoveOptions` at all. (You can always use `typeof moveOptionsGetters`). But, it's not really hurting anything.

**Review by d0sboots — COMMENTED** [review]
Great! That cleaned up a lot.

**Comment by d0sboots** [comment]
BTW this is a PR on catlover's PR, not the base repository, so I can't merge it. I didn't realize that until now; that's a little awkward because the base PR has already been merged. 

If you can rebase this to be off of dev instead, that would be ideal, but it will require some advanced git-foo. Otherwise, you might want to open a new PR.

**Comment by catloversg** [comment]
When you feel that it's ready to merge, please mark this PR as "Approved". After that, I'll close this PR and do all necessary Git operations to open a new PR with this branch's commits on the main repository.

**Review by d0sboots — APPROVED** [review]
Approval

**Pull request closed without merging**
